content/blog/2024/10/what-is-the-esp-registry/index.md

```diff
@@ -0,0 +1,147 @@
+---
+title: "What is the ESP-Registry?"
+date: 2024-10-27
+showAuthor: false
+authors:
+  - "pedro-minatel"
+tags: ["Registry", "Component", "ESP-IDF"]
+---
+
+When starting a new project, we often reuse code from existing components, such as libraries and drivers, to save time. This not only saves time but also makes your project easier to maintain.
+
+In this article, we will explore what the ESP-Registry is and the key concepts of components, making it easy to understand not only how to use components but also how to contribute.
+
+## Concept
+
+In the [ESP-IDF](https://idf.espressif.com), components are widely used to make the framework more modular, making it clearer to understand and maintain. The naming convention in the ESP-IDF is to use the term "component" instead of "library" because the ESP-IDF is more like an operating system with self-contained modules that interact with the system at a higher level.
+
+There are three concepts that we need to introduce in order to understand the complete solution:
+
+### 1. Component
+
+As mentioned before, the component is a self-contained piece of code that can be used to add a new functionality to the application. In most of the cases, the component is used in several applications, making easy to reuse code and as well will improve the efficiency and reduce the development complexity.
+
+For example, in your application (`my-application`), you are going to use a specific sensor (`sensor-component`) that will collect temperature and humidity from the environment. Instead of adding all the logic behind the communication protocol, data collection, conversions and etc to the application code, you can add as a component.
+
+```text
+.
+└── my-application
+    ├── components
+    │   └── sensor-component
+    │       └── include
+    │           └── sensor-component.c
+    │       └── src
+    │           └── CMakeLists.txt
+    │           └── sensor-component.c
+    │       └── CMakeLists.txt
+    ├── CMakeLists.txt

```

Comment:
This application layout is not a typical one, usually we have a `main` component where the code is contained:

```
    ├── CMakeLists.txt
    ├── main
    │   └── CMakeLists.txt
    │   └── my-application.c
    |   └── include
    │       └── my-application.h
    └── my-application.c
```

Reply:
One more thing. Should we create the `src` folder for the component?
Currently, the `idf.py create-component` does not create the `scr` folder.

Reply:
I think 'src' directory is optional, most simple components inside ESP-IDFs components directory do not have it and source files are directly in the component directory.

However, I want to update "create-component", "create-project" and 'create-manifest' templates because they are too minimalistic and a bit outdated now.

Reply:
I'll remove this part from this article and add it to the new article "Creating Components".